### Tugas 2
- Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv
2. Fitur CountVectorizer dengan mengaktifkan stop_words
3. Evaluasi hasilnya

- Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur TF-IDF dengan mengaktifkan stop_words

3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

Pra Pengolahan Data

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, classification_report

# 1. Load Data dan Pra-pengolahan
df = pd.read_csv('/content/drive/MyDrive/Datasets/spam.csv', encoding='latin-1')
df = df.iloc[:, [0, 1]]
df.columns = ['label', 'message']

# Encoding label 'ham'/'spam' menjadi 0/1
le = LabelEncoder()
df['label_encoded'] = le.fit_transform(df['label'])

X_text = df['message']
y = df['label_encoded']

# Pembagian data (Train 75%, Test 25%)
X_train_text, X_test_text, y_train, y_test = train_test_split(
    X_text, y, test_size=0.25, random_state=42, stratify=y
)

results = {}

print("Data pesan telah dimuat, di-encode, dan dibagi menjadi set latih dan uji.")

Data pesan telah dimuat, di-encode, dan dibagi menjadi set latih dan uji.


**Model Klasifikasi MNB dengan CountVectorizer**

Tujuan: Mengklasifikasikan SMS dengan menggunakan model Multinomial Naive Bayes (MNB), mengekstrak fitur berdasarkan frekuensi kata muncul.

In [ ]:
# TUGAS 1: Multinomial Naive Bayes (CountVectorizer)

# Langkah 1.1: Feature Extraction (CountVectorizer)
# Mengubah teks menjadi matriks hitungan kata. stop_words='english' menghilangkan kata umum.
cv = CountVectorizer(stop_words='english')
X_train_cv = cv.fit_transform(X_train_text)
X_test_cv = cv.transform(X_test_text)

# Langkah 1.2: Pelatihan Model
mnb_cv = MultinomialNB()
mnb_cv.fit(X_train_cv, y_train)

# Langkah 1.3: Evaluasi Hasil
y_pred_cv = mnb_cv.predict(X_test_cv)
acc_cv = accuracy_score(y_test, y_pred_cv)

print("--- Hasil CountVectorizer ---")
print(f"Akurasi: {acc_cv:.4f}")
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_cv, target_names=le.classes_))

--- Hasil CountVectorizer ---
Akurasi: 0.9864
Laporan Klasifikasi:
               precision    recall  f1-score   support

         ham       0.99      0.99      0.99      1206
        spam       0.96      0.94      0.95       187

    accuracy                           0.99      1393
   macro avg       0.98      0.97      0.97      1393
weighted avg       0.99      0.99      0.99      1393



**Model Klasifikasi MNB dengan TF-IDF**

Tujuan: Mengklasifikasikan SMS menggunakan model MNB, dengan fitur yang diekstrak berdasarkan pentingnya kata (TF-IDF).

In [ ]:
# TUGAS 2: Multinomial Naive Bayes (TF-IDF)

print("\n--- TUGAS 2: TF-IDF ---")

# Langkah 2.1: Feature Extraction (TF-IDF Vectorizer)
# Mengubah teks menjadi matriks yang mencerminkan "pentingnya" kata.
tfidf = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf.fit_transform(X_train_text)
X_test_tfidf = tfidf.transform(X_test_text)

# Langkah 2.2: Pelatihan Model
mnb_tfidf = MultinomialNB()
mnb_tfidf.fit(X_train_tfidf, y_train)

# Langkah 2.3: Evaluasi Hasil
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)

print(f"Akurasi: {acc_tfidf:.4f}")
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_tfidf, target_names=le.classes_))


--- TUGAS 2: TF-IDF ---
Akurasi: 0.9691
Laporan Klasifikasi:
               precision    recall  f1-score   support

         ham       0.97      1.00      0.98      1206
        spam       1.00      0.77      0.87       187

    accuracy                           0.97      1393
   macro avg       0.98      0.89      0.93      1393
weighted avg       0.97      0.97      0.97      1393



**Kesimpulan dan Perbandingan Fitur Terbaik**

Langkah terakhir membandingkan metrik kunci dari kedua model.

In [ ]:
# 4. Kesimpulan dan Perbandingan Metrik
acc_cv = 0.9864
acc_tfidf = 0.9691
precision_spam_cv = 0.9615
recall_spam_cv = 0.9358
f1_spam_cv = 0.9485
precision_spam_tfidf = 1.0000
recall_spam_tfidf = 0.7701
f1_spam_tfidf = 0.8701

comparison_table = pd.DataFrame({
    'Metrik': ['Akurasi', 'Presisi Spam', 'Recall Spam', 'F1-Score Spam'],
    'CountVectorizer': [acc_cv, precision_spam_cv, recall_spam_cv, f1_spam_cv],
    'TF-IDF': [acc_tfidf, precision_spam_tfidf, recall_spam_tfidf, f1_spam_tfidf]
})

print("\n--- PERBANDINGAN FITUR (CV vs TF-IDF) ---")
print(comparison_table.to_markdown(index=False, floatfmt=".4f"))


--- PERBANDINGAN FITUR (CV vs TF-IDF) ---
| Metrik        |   CountVectorizer |   TF-IDF |
|:--------------|------------------:|---------:|
| Akurasi       |            0.9864 |   0.9691 |
| Presisi Spam  |            0.9615 |   1.0000 |
| Recall Spam   |            0.9358 |   0.7701 |
| F1-Score Spam |            0.9485 |   0.8701 |


####Kesimpulan
1. CountVectorizer adalah fitur terbaik yang akurat dan keseimbangan tinggi. CountVectorizer unggul dalam akurasi keseluruhan ($0.9864$) dan F1-Score ($0.9485$) untuk kelas "spam" karena menyeimbangkan Presisi dan Recall. F1-Score juga adalah metrik terbaik untuk menilai kinerja pada kelas minoritas, atau "spam", karena menyeimbangkan Presisi dan Recall.


 2. Recall yang Lebih Penting: Dalam filter spam, kemampuan model untuk mengidentifikasi semua pesan spam yang ada merupakan faktor yang sangat penting.   Untuk Recall, CountVectorizer  $0.94, jauh lebih banyak daripada TF-IDF ($0.77), yang menunjukkan bahwa CountVectorizer lebih baik dalam mencegah pesan spam masuk ke kotak masuk.

 - Oleh karena itu, meskipun TF-IDF 1.00 untuk Presisi Sempurna, model tersebut kurang efektif dibandingkan CountVectorizer dalam aplikasi filter spam yang sebenarnya.